### PR TITLE
8313357: Revisit requiring SA tests on OSX to either run as root or use sudo

### PR DESCRIPTION
**Pull request by plummercj**
On OSX, don't require that sudo be used to launch SA tools if developer mode is enabled. More details are in the CR. 

Due to this change, the following tests are no longer being skipped if the host has developer mode is enabled. They previously required running as root because if sudo was used some files were created with root ownership and could not be deleted:

serviceability/sa/ClhsdbDumpclass.java
serviceability/sa/TestClassDump.java

I tested on my home system (M1) with developer mode enabled and also disabled. Also tested with CI, which appears to have developer mode disabled on most hosts, but there are a few where it is enabled.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313357](https://bugs.openjdk.org/browse/JDK-8313357): Revisit requiring SA tests on OSX to either run as root or use sudo (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15238/head:pull/15238` \
`$ git checkout pull/15238`

Update a local copy of the PR: \
`$ git checkout pull/15238` \
`$ git pull https://git.openjdk.org/jdk.git pull/15238/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15238`

View PR using the GUI difftool: \
`$ git pr show -t 15238`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15238.diff">https://git.openjdk.org/jdk/pull/15238.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15238#issuecomment-1674138317)

### CODE REVIEW

File: test/lib/jdk/test/lib/SA/SATestUtils.java [hunk 1]
```diff
@@ -77,11 +78,54 @@ public static void skipIfCannotAttach() {
...
+            try {
+                p.waitFor();
+            } catch (InterruptedException e) {
+                throw new RuntimeException("DevToolsSecurity process interrupted", e);
+            }
+
+            String out = new String(p.getInputStream().readAllBytes());
+            String err = new String(p.getErrorStream().readAllBytes());
+            System.out.print("DevToolsSecurity stdout: " + out);
+            if (out.equals("")) System.out.println();
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/SA/SATestUtils.java:113]: This looks odd - why not just an unconditional println?

File: test/lib/jdk/test/lib/SA/SATestUtils.java [hunk 2]
```diff
@@ -77,11 +78,54 @@ public static void skipIfCannotAttach() {
...
+        }
+        return !developerModeEnabled();
     }
 
+    /*
+     * Run "DevToolsSecurity --status" to see if developer mode is enabled.
+     */
+    private static boolean developerModeEnabled() {
+        List<String> cmd = new ArrayList<String>();
+        cmd.add("DevToolsSecurity");
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/SA/SATestUtils.java:98]: Is this tool likely to be in the path?

**Review by dholmes-ora — COMMENTED** [review]
This seems reasonable in principle but I would be concerned about the overhead of exec'ing another process here. How many test actually use this? Is this query only ever executed per VM lifetime? (otherwise we should cache the result).

Thanks.

File: test/lib/jdk/test/lib/SA/SATestUtils.java (hunk 1 above)
Comment (plummercj) [reply test/lib/jdk/test/lib/SA/SATestUtils.java:113]: If it is not "", then `out` will include a newline at the end of the last line, which means you'd end up with an empty line in the output. I considered the following:

            if (!out.equals("")) {
                System.out.print("DevToolsSecurity stdout: " + out);
            }

But it prints nothing in the output if it is "", and I that's not what I was looking for.

File: test/lib/jdk/test/lib/SA/SATestUtils.java (hunk 2 above)
Comment (plummercj) [reply test/lib/jdk/test/lib/SA/SATestUtils.java:98]: It is for my work machine, my home machine, and the CI machines. If not installed then an exception is thrown. It probably requires that xcode be installed, but this is already a requirement to run the tests.

**Comment by plummercj** [comment]
> This seems reasonable in principle but I would be concerned about the overhead of exec'ing another process here. How many test actually use this? Is this query only ever executed per VM lifetime? (otherwise we should cache the result).

Every SA test uses it except I think one test that doesn't need any special privileges to run. That's about 65 tests.

This PR actually reduces the number of processes needed *IF* developer mode is enabled. Previously launching an SA tool required 3 processes: one to first check if passwordless sudo works, one to launch the SA tool with sudo, and one for the SA tool itself. So now there are 4 *IF* developer mode is not enabled (added a developer mode check), but there are only two if developer mode is enabled (developer mode check and SA tool launch). 

BTW, one thing I didn't really call out in the description is that this allows the SA tests to be run when passwordless sudo is not enabled but developer mode is. Previously they would have just all thrown SkippedException.

I'm not sure about your caching question. I would assume you are asking how many times it executed per test, which is only once. I can't see caching between test runs because the setting can be changed by the user.

**Comment by dholmes-ora** [comment]
> I'm not sure about your caching question. I would assume you are asking how many times it executed per test, which is only once.

Okay. That is still unfortunate as I would expect this to be a dynamic property that changes very rarely.

> I can't see caching between test runs because the setting can be changed by the user.

I would not be worried about trying to deal with a user that changes the setting in that way. Test environments are expected to kept in a stable condition.

**Comment by dholmes-ora** [comment]
> but there are only two if developer mode is enabled (developer mode check and SA tool launch).

So do we expect developer mode to be enabled in our CI testing?

**Comment by plummercj** [comment]
> So do we expect developer mode to be enabled in our CI testing?

We don't expect it to be, but it is on a few machines.

**Comment by plummercj** [comment]
> I would not be worried about trying to deal with a user that changes the setting in that way. Test environments are expected to kept in a stable condition.

I think it is best to test for developer mode on each test run. The user might change it after a test run after seeing SkippedException, especially if they are aware that developer mode will address it. They may normally run with it disabled, but want it enabled when running SA tests.

**Review by dholmes-ora — APPROVED** [review]
Thumbs up from me.

Thanks

**Comment by plummercj** [comment]
Ping! I could use one more review please. Thanks

**Review by alexmenkov — APPROVED** [review]

**Comment by plummercj** [comment]
Thank you Alex and David!

/integrate

**Pull request closed without merging**